packages/p2p/src/components/modal-manager/modal-manager-context-provider.tsx

```diff
@@ -1,15 +1,54 @@
 import React from 'react';
-import { useStores } from 'Stores';
+import { useStores } from 'Stores/index';
 import { ModalManagerContext } from './modal-manager-context';
 import { isDesktop } from '@deriv/shared';
+import type { TModalKeys, TModalProps } from 'Constants/modals';
 
-const ModalManagerContextProvider = props => {
-    const [active_modal, setActiveModal] = React.useState({});
-    const [previous_modal, setPreviousModal] = React.useState({});
+/**
+ * TModal type represents a specific modal type, with its prop typing constrained to that specific modal based on its key.
+ */
+export type TModal<T extends TModalKeys> = {
+    key: T;
+    props: TModalProps[T];
+};
+
+/**
+ * TModalVariants type represents the union of all possible modal keys and prop types.
+ * This type is necessary for typing a function or generic argument that could accept any type of modal.
+ */
+type TModalVariants = {
+    key: TModalKeys;
+    props: TModalProps[TModalKeys];
+};
+
+export type TModalManagerContext = {
+    hideModal: (options?: THideModalOptions) => void;
+    is_modal_open: boolean;
+    isCurrentModal: (...keys: TModalKeys[]) => boolean;
+    modal_props: Map<TModalKeys, TModalProps[TModalKeys]>;
+    modal: TModalVariants | null;
+    previous_modal: TModalVariants | null;
+    showModal: <T extends TModalKeys>(modal: TModal<T>, options?: TShowModalOptions) => void;
+    stacked_modal: TModalVariants | null;
+    useRegisterModalProps: <T extends TModalKeys>(modals: TModal<T> | TModalVariants[]) => void;
+};
+
+type TShowModalOptions = {
+    should_stack_modal?: boolean;
+};
+
+type THideModalOptions = {
+    should_save_form_history?: boolean;
+    should_hide_all_modals?: boolean;
+};
+
+const ModalManagerContextProvider = (props: React.PropsWithChildren<{ mock?: TModalManagerContext }>) => {
+    const [active_modal, setActiveModal] = React.useState<TModalVariants | null>(null);
+    const [previous_modal, setPreviousModal] = React.useState<TModalVariants | null>(null);
```

Comment:
```suggestion
    const [modal, setModal] = React.useState<{ active: TModalVariants, previous: TModalVariants }>(null);
```
Looks better combined. You could move this to a hook and abstract out setting current and previous modals. Ideally, the user does not need to know how to set the previous modal